### PR TITLE
User menu was updated (implements base version)

### PR DESCRIPTION
**Pull request by vshovkuta**
+ added the user-menu.component (for authorizated users);
+ added the separated buttons for 'Sing in'/'Sing up'
+ added the permission labels for the navbar's link
^ updated the guard.service
^ updated the path (added the canActivate properties)
^ updated the authentication.service (added redirect to the "Find Shelter" page after successful authorization)
^ updated the styles for the navbar
[del] ^ updated tthe permissions.enum and permissions.service (added navigate* items) [del] 

pr0
pr1

### CODE REVIEW

File: src/app/navbar/user-menu/user-menu.component.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  <a mat-button [matMenuTriggerFor]="menu">
+    {{authStateService.getStateProperty('email')}}
+    <mat-icon>arrow_drop_down</mat-icon>
+  </a>
+
+  <mat-menu #menu="matMenu">
+    <a *appPermission="permissions.navigateRegistrateShelter" mat-menu-item [routerLink]="['/registration-shelter']">
+      Registrate a New Shelter</a>
+    <a *appPermission="permissions.navigateUsersDetails" mat-menu-item [routerLink]="['/users']">Users Details</a>
+    <a *appPermission="permissions.navigateLogout" mat-menu-item (click)="authenticationService.logout()">Sign out</a>
```
Comment (ksenya-ta) [inline src/app/navbar/user-menu/user-menu.component.html]: how is it possible that user won't have permissions to perform logout???

File: src/app/navbar/user-menu/user-menu.component.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  </mat-menu>
+</div>
+
+<ng-template #guest>
+  <a *appPermission="permissions.navigateLogin" mat-button [routerLink]="['/login']">Sign in</a>
+  <a *appPermission="permissions.navigateRegistrateUser" mat-raised-button [routerLink]="['/registraction-user']">
```
Comment (ksenya-ta) [inline src/app/navbar/user-menu/user-menu.component.html]: unauthorized should not have separate set of permissions

File: src/app/navbar/user-menu/user-menu.component.html [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<div *ngIf="authStateService.isLogged(); else guest">
+  <a mat-button [matMenuTriggerFor]="menu">
+    {{authStateService.getStateProperty('email')}}
+    <mat-icon>arrow_drop_down</mat-icon>
+  </a>
+
+  <mat-menu #menu="matMenu">
+    <a *appPermission="permissions.navigateRegistrateShelter" mat-menu-item [routerLink]="['/registration-shelter']">
```
Comment (ksenya-ta) [inline src/app/navbar/user-menu/user-menu.component.html]: do not create separate permissions for navigation. Permissions should be based on the performed actions. If user don't have permission to register shelters it is clear that he should not see the link to open this page. And the same for all other menu items

File: src/app/navbar/user-menu/user-menu.component.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  selector: 'app-user-menu',
+  templateUrl: './user-menu.component.html',
+  styleUrls: ['./user-menu.component.scss']
+})
+export class UserMenuComponent {
+  public permissions = Permissions;
+
+  constructor(
+    private authStateService: AuthStateService,
+    private authenticationService: AuthenticationService
```
Comment (ksenya-ta) [inline src/app/navbar/user-menu/user-menu.component.ts]: do not add dependencies you're not using

File: src/app/shared/permissions/models/permissions.enum.ts [hunk 5]
```diff
@@ -3,5 +3,14 @@ export enum Permissions {
...
+  shelterUploadPhoto,
+  navigateDonate,
+  navigateAuction,
+  navigateShelters,
+  navigateLogo,
+  navigateRegistrateShelter,
+  navigateRegistrateUser,
+  navigateUsersDetails,
+  navigateLogin,
+  navigateLogout
```
Comment (ksenya-ta) [inline src/app/shared/permissions/models/permissions.enum.ts]: do not add separate permissions for navigation, use the ones was already added for appropriate CRUD operations, if user don't have Read permission you should not render the links for it in menu

File: src/app/shared/services/user/authentication.service.ts [hunk 6]
```diff
@@ -27,12 +29,17 @@ export class AuthenticationService {
       map(stateObj => {
```
Comment (ksenya-ta) [inline src/app/shared/services/user/authentication.service.ts]: this kind of actions should not be part of `map`. only mapping the data can be added in map. if you need to perform something once you get the data you better look into the `tap` operator instead

File: src/app/navbar/user-menu/user-menu.component.ts (hunk 4 above)
Comment (vshovkuta) [reply src/app/navbar/user-menu/user-menu.component.ts]: They is used

File: src/app/app-routing.module.ts [hunk 7]
```diff
@@ -3,12 +3,19 @@ import { Routes, RouterModule } from '@angular/router';
...
   { path: '', redirectTo: '/shelters', pathMatch: 'full'},
   { path: 'login', component: LoginComponent },
   { path: 'registraction-user', component: RegistrationUserComponent },
-  { path: 'users', component: AdminUsersComponent }
+  { path: 'users', component: AdminUsersComponent,
+    canActivate: [GuardService],
+    data: { allowedRoles: ['Admin'] } },
+  { path: 'registration-shelter', component: ShelterRegistrationComponent,
+    canActivate: [GuardService],
+    data: { allowedRoles: ['Admin'] } }
```
Comment (ksenya-ta) [inline src/app/app-routing.module.ts]: please format it properly

File: src/app/app-routing.module.ts [hunk 8]
```diff
@@ -3,12 +3,19 @@ import { Routes, RouterModule } from '@angular/router';
...
+import { GuardService } from './shared/guard.service';
 
 const routes: Routes = [
   { path: '', redirectTo: '/shelters', pathMatch: 'full'},
   { path: 'login', component: LoginComponent },
   { path: 'registraction-user', component: RegistrationUserComponent },
-  { path: 'users', component: AdminUsersComponent }
+  { path: 'users', component: AdminUsersComponent,
+    canActivate: [GuardService],
+    data: { allowedRoles: ['Admin'] } },
```
Comment (ksenya-ta) [inline src/app/app-routing.module.ts]: use enum for roles

File: src/app/navbar/user-menu/user-menu.component.html [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<div *ngIf="authStateService.isLogged(); else guest">
+  <a mat-button [matMenuTriggerFor]="menu">
+    {{authStateService.getStateProperty('email')}}
+    <mat-icon>arrow_drop_down</mat-icon>
+  </a>
+
+  <mat-menu #menu="matMenu">
+    <a *ngIf="authStateService.checkRoles(['Admin'])" mat-menu-item [routerLink]="['/registration-shelter']">
```
Comment (ksenya-ta) [inline src/app/navbar/user-menu/user-menu.component.html]: you have directive that checks permissions, use it instead

File: src/app/navbar/user-menu/user-menu.component.ts (hunk 4 above)
Comment (ksenya-ta) [reply src/app/navbar/user-menu/user-menu.component.ts]: where? they are private and there are no methods in this component (btw do not use service directly from template and do not use private methods directly from template)

File: src/app/shared/guard.service.ts [hunk 10]
```diff
@@ -1,23 +1,30 @@
...
+  ): boolean {
+    if (this.authStateService.checkRoles(route.data.allowedRoles)) {
       return true;
-    }
-
-    this.router.navigate(['/login'], {
-      queryParams: {
-        auth: false
+    } else {
+      this.notifierService.showNotice('You can\'t use this path directly!', 'warning');
```
Comment (ksenya-ta) [inline src/app/shared/guard.service.ts]: please add more user friendly messages

File: src/app/shared/guard.service.ts [hunk 11]
```diff
@@ -1,23 +1,30 @@
...
+      if (this.authStateService.isLogged()) {
+        this.router.navigate(['']);
+      } else {
+        this.router.navigate(['/login']);
       }
```
Comment (ksenya-ta) [inline src/app/shared/guard.service.ts]: why do we need to redirect to logic/home page? it should be ok to stay on the previous page

File: src/app/shared/services/state/auth-state.service.ts [hunk 12]
```diff
@@ -20,6 +20,10 @@ export class AuthStateService {
     return Boolean(this.getStateValue());
   }
 
+  public checkRoles(array: string[]): boolean {
```
Comment (ksenya-ta) [inline src/app/shared/services/state/auth-state.service.ts]: name this properly, do not use reserved words for naming (please refer to this list to review what you should avoid https://www.w3schools.com/js/js_reserved.asp)

File: src/app/shared/validators/form-fields-validator.ts [hunk 13]
```diff
@@ -20,7 +20,7 @@ export class FormFiledsValidator {
       '^([a-zA-Z0-9_.-])+@(([a-zA-Z0-9-])+[.])+([a-zA-Z0-9]{2,4})+$'
     ),
     Validators.minLength(6),
-    Validators.maxLength(20)
+    Validators.maxLength(50)
```
Comment (ksenya-ta) [inline src/app/shared/validators/form-fields-validator.ts]: if you change validation rules you should also update it in the hints on the appropriate pages

File: src/app/shared/guard.service.ts (hunk 10 above)
Comment (vshovkuta) [reply src/app/shared/guard.service.ts]: ✓ done

File: src/app/shared/services/state/auth-state.service.ts (hunk 12 above)
Comment (vshovkuta) [reply src/app/shared/services/state/auth-state.service.ts]: Renamed to 'arrayOfRoles'
✓ done

File: src/app/shared/guard.service.ts (hunk 11 above)
Comment (vshovkuta) [reply src/app/shared/guard.service.ts]: ~ Left a previous version

File: src/app/navbar/navbar.component.scss [hunk 14]
```diff
@@ -29,4 +30,9 @@
   .flex-spacer {
     flex-grow: 1;
   }
+
+  app-user-menu {
+    display: flex;
+    justify-content: space-between;
+  }
```
Comment (ksenya-ta) [inline src/app/navbar/navbar.component.scss]: this should be part of app-user-menu styles

File: src/app/shared/services/state/auth-state.service.ts [hunk 15]
```diff
@@ -20,6 +20,10 @@ export class AuthStateService {
     return Boolean(this.getStateValue());
   }
 
+  public checkRoles(arrayOfRoles: string[]): boolean {
+    return arrayOfRoles.some((item) => item === this.getStateProperty('roles')[0]);
```
Comment (ksenya-ta) [inline src/app/shared/services/state/auth-state.service.ts]: do not call `getStateProperty `in a loop, get the value once and then compare with it

**Pull request merged**